v3/mathjax3-ts/core/MathDocument.ts

```diff
@@ -0,0 +1,401 @@
+
+    /*
+     * The options for the document
+     */
+    options: OptionList;
+
+    /*
+     * The list of MathItems found in this page
+     */
+    math: MathList;
+
+    /*
+     * This object tracks what operations have been performed, so that (when
+     *  asynchronous operations are used), the ones that have already been
+     *  completed won't be performed again.
+     */
+    processed: {[name: string]: boolean};
+
+    /*
+     * An array of input jax to run on the document
+     */
+    inputJax: InputJax[];
+
+    /*
+     * The output jax to use for the document
+     */
+    outputJax: OutputJax;
+
+    /*
+     * Locates the math in the document and constructs the MathList
+     *  for the document.
+     *
+     * @param{OptionList} options  The options for locating the math
+     * @return{MathDocument}       The math document instance
+     */
+    findMath(options?: OptionList): MathDocument;
+
+    /*
+     * Calls the input jax to process the MathItems in the MathList
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    compile(): MathDocument;
+
+    /*
+     * Gets the metric information for the MathItems
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    getMetrics(): MathDocument;
+
+    /*
+     * Calls the output jax to process the compiled math in the MathList
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    typeset(): MathDocument;
+
+    /*
+     * Add any event handlers to the typeset math
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    addEventHandlers(): MathDocument;
+
+    /*
+     * Updates the document to include the typeset math
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    updateDocument(): MathDocument;
+
+    /*
+     * Removes the typeset math from the document
+     *
+     * @param{boolean} restore  True if the original math should be put
+     *                            back into the document as well
+     * @return{MathDocument}    The math document instance
+     */
+    removeFromDocument(restore?: boolean): MathDocument;
+
+    /*
+     * Set the state of the document (allowing you to roll back
+     *  the state to a previous one, if needed).
+     *
+     * @param{boolean} restore  True if the original math should be put
+     *                            back into the document during the rollback
+     * @return{MathDocument}    The math document instance
+     */
+    state(state: number, restore?: boolean): MathDocument;
+
+    /*
+     * Clear the processed values so that the document can be reprocessed
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    reset(): MathDocument;
+
+    /*
+     * Reset the processed values and clear the MathList (so that new math
+     * can be processed in the document).
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    clear(): MathDocument;
```

Comment:
`removeFromDocument()` takes the math back out of the document, putting back the original math.  `clear()` clears the processed status indicators so that you can process new math added to the document (without losing track of the math already in the document, and leaving the typeset math in the document).  `reset()` clears the math list as well as the processed status, i.e., forgets everything it knows about the page, but leaves the typeset math in place.